testbed/sampling/sampling_data_provider.go

```diff
@@ -0,0 +1,176 @@
+	// Total traces generated
+	tracesGenerated int
+	// we will use this to ensure that the received spans matches the expected sampled spans
+	sampledSpansGenerated atomic.Uint64
+	// Controls if a trace should be sampled
+	shouldSample shouldSampleFunc
+	// Customize spans that need to be sampled so that they can properly filtered
+	sampledSpanCustomizer sampledSpanCustomizerFunc
+	// Stores all the trances that were sampled
+	expectedSampledTraces []ptrace.Traces
+}
+
+type sampledSpanCustomizerFunc func(span ptrace.Span)
+
+type shouldSampleFunc func(traceIDSequence uint64) bool
+
+// SamplingDataProviderOption defines a PollutedDataProvider option.
+type SamplingDataProviderOption func(t *samplingDataProvider)
+
+// WithShouldSampleFunction allows you to control if a trace should be sampled.
+func WithSamplingFunction(shouldSample shouldSampleFunc) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.shouldSample = shouldSample
+	}
+}
+
+// WithSampledSpanCustomizer allows you to customize the spans that are sampled.
+func WithSampledSpanCustomizer(customizer sampledSpanCustomizerFunc) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.sampledSpanCustomizer = customizer
+	}
+}
+
+// WithExpectedSampledTraces allows you to specify the number of spans per trace
+func WithSpansPerTrace(spansPerTrace int) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.spansPerTrace = spansPerTrace
+	}
+}
+
+// WithExpectedSampledTraces allows you to specify the total number of traces generated in a run
+func WithNumTracesToGenerate(numTracesToGenerate int) SamplingDataProviderOption {
+	return func(tc *samplingDataProvider) {
+		tc.numTracesToGenerate = numTracesToGenerate
+	}
+}
+
+// NewSamplingDataProvider creates a new instance of samplingDataProvider
+func NewSamplingDataProvider(opts ...SamplingDataProviderOption) *samplingDataProvider {
+	sdp := samplingDataProvider{
+		// Defaults. These are arbitrary values.
+		spansPerTrace:       6,
+		numTracesToGenerate: 50,
+	}
+
+	for _, opt := range opts {
+		opt(&sdp)
+	}
+
+	return &sdp
+}
+
+// SetLoadGeneratorCounters sets the data providers items generated field to have a ref to the load generators counter
+func (sdp *samplingDataProvider) SetLoadGeneratorCounters(dataItemsGenerated *atomic.Uint64) {
+	sdp.dataItemsGenerated = dataItemsGenerated
+}
+
+// GenerateTraces the boolean in this return reflects whether we are done generating traces or not.
+func (sdp *samplingDataProvider) GenerateTraces() (ptrace.Traces, bool) {
+	if sdp.tracesGenerated >= sdp.numTracesToGenerate {
+		return ptrace.NewTraces(), true
+	}
+	sdp.tracesGenerated += 1
+
+	traceData := sdp.generateSampledTrace()
+
+	return traceData, false
+
+}
+
+func (sdp *samplingDataProvider) ToBeSampledSpansGenerated() uint64 {
+	return sdp.sampledSpansGenerated.Load()
+}
+
+func (sdp *samplingDataProvider) generateSampledTrace() ptrace.Traces {
```

Comment:
```suggestion
func (sdp *samplingDataProvider) generateTraceData() ptrace.Traces {
```
This now does more than generating sampled traces.